sql/postgresql/PostgreSQLParser.g4

```diff
@@ -503,32 +503,19 @@ discardstmt
    ;
 
 altertablestmt
-   : ALTER TABLE relation_expr alter_table_cmds
-   | ALTER TABLE IF_P EXISTS relation_expr alter_table_cmds
-   | ALTER TABLE relation_expr partition_cmd
-   | ALTER TABLE IF_P EXISTS relation_expr partition_cmd
-   | ALTER TABLE ALL IN_P TABLESPACE name SET TABLESPACE name opt_nowait
-   | ALTER TABLE ALL IN_P TABLESPACE name OWNED BY role_list SET TABLESPACE name opt_nowait
-   | ALTER INDEX qualified_name alter_table_cmds
-   | ALTER INDEX IF_P EXISTS qualified_name alter_table_cmds
-   | ALTER INDEX qualified_name index_partition_cmd
-   | ALTER INDEX ALL IN_P TABLESPACE name SET TABLESPACE name opt_nowait
-   | ALTER INDEX ALL IN_P TABLESPACE name OWNED BY role_list SET TABLESPACE name opt_nowait
-   | ALTER SEQUENCE qualified_name alter_table_cmds
-   | ALTER SEQUENCE IF_P EXISTS qualified_name alter_table_cmds
-   | ALTER VIEW qualified_name alter_table_cmds
-   | ALTER VIEW IF_P EXISTS qualified_name alter_table_cmds
-   | ALTER MATERIALIZED VIEW qualified_name alter_table_cmds
-   | ALTER MATERIALIZED VIEW IF_P EXISTS qualified_name alter_table_cmds
-   | ALTER MATERIALIZED VIEW ALL IN_P TABLESPACE name SET TABLESPACE name opt_nowait
-   | ALTER MATERIALIZED VIEW ALL IN_P TABLESPACE name OWNED BY role_list SET TABLESPACE name opt_nowait
-   | ALTER FOREIGN TABLE relation_expr alter_table_cmds
-   | ALTER FOREIGN TABLE IF_P EXISTS relation_expr alter_table_cmds
+   : ALTER TABLE (IF_P EXISTS)? relation_expr (alter_table_cmds | partition_cmd)
+   | ALTER TABLE ALL IN_P TABLESPACE name (OWNED BY role_list)? SET TABLESPACE name opt_nowait
+   | ALTER INDEX (IF_P EXISTS)? qualified_name (alter_table_cmds | index_partition_cmd)
+   | ALTER INDEX ALL IN_P TABLESPACE name (OWNED BY role_list)? SET TABLESPACE name opt_nowait
+   | ALTER SEQUENCE (IF_P EXISTS)? qualified_name alter_table_cmds
+   | ALTER VIEW (IF_P EXISTS)? qualified_name alter_table_cmds
+   | ALTER MATERIALIZED VIEW (IF_P EXISTS)? qualified_name alter_table_cmds
+   | ALTER MATERIALIZED VIEW ALL IN_P TABLESPACE name (OWNED BY role_list)? SET TABLESPACE name opt_nowait
+   | ALTER FOREIGN TABLE (IF_P EXISTS)? relation_expr alter_table_cmds
```

Comment:
I suggest extracting the common part:

```suggestion
   : ALTER ( TABLE ( (IF_P EXISTS)? relation_expr (alter_table_cmds | partition_cmd)
                   | ALL IN_P TABLESPACE name (OWNED BY role_list)? SET TABLESPACE name opt_nowait )
           | INDEX ( (IF_P EXISTS)? qualified_name (alter_table_cmds | index_partition_cmd)
                   | ALL IN_P TABLESPACE name (OWNED BY role_list)? SET TABLESPACE name opt_nowait )
           | SEQUENCE (IF_P EXISTS)? qualified_name alter_table_cmds
           | VIEW (IF_P EXISTS)? qualified_name alter_table_cmds
           | MATERIALIZED VIEW ( (IF_P EXISTS)? qualified_name alter_table_cmds
                               | ALL IN_P TABLESPACE name (OWNED BY role_list)? SET TABLESPACE name opt_nowait )
           | FOREIGN TABLE (IF_P EXISTS)? relation_expr alter_table_cmds )
```

Reply:
I did not extract common parts (yet) for not so frequently used statements, cause it does not significantly hurts performance
also I plan to use https://github.com/kaby76/Domemtech.Trash to make some cleanup/changes to grammar - learning now how to use it properly